File: backend/Controllers/userController.js
const userModel = require("../Models/userModel.js")
const bcrypt = require("bcryptjs")
const jwt = require("jsonwebtoken")


exports.signup = async (req, res) => {
    try {

        const user = await userModel.findOne({ email: req.body?.email });

        if ((user) && (user.active == false)) {

            const token = jwt.sign({ _id: user._id }, process.env.TOKEN_SECRET_KEY)

            const hashedPassword = await bcrypt.hash(req.body.pass, 10);

            await userModel.updateOne({ email: user.email }, { $set: { fname: req.body.fname, lname: req.body.lname, email: req.body.email, pass: hashedPassword, active: true, token: token } }, { new: true })
            //Update the data for the already existing email ID whose account has been deleted.

            res.status(201).json({
                status: true,
                message: "Signup successfully..!",
                data: token,
            });
        }
        else if (user !== null) {
            return res.status(400).json({
                status: false,
                message: "User already exist..!",
            });
        }
        else {
            if (req.body.pass === req.body.cpass) {

                const data = new userModel({
                    fname: req.body.fname,
                    lname: req.body.lname,
                    email: req.body.email,
                    mno: req.body.mno,
                    pass: req.body.pass
                })

                const token = await data.gentoken();


                await data.save();  // password will be bcrypted using function which is define in userModel

                await userModel.updateOne({ email: req.body.email }, { $set: { token: token } }, { new: true });
                //first save and then update


                res.status(201).json({
                    status: true,
                    message: "Signup successfully..!",
                    data: token,
                });

            } else {
                res.status(400).json({
                    status: false,
                    message: "Password and confirm password does not match..!",
                });

            }
        }

    } catch (error) {
        res.status(400).json({
            status: false,
            message: error.message,
        });
    }
}


exports.signin = async (req, res) => {
    try {
        const data = await userModel.findOne({ email: req.body.email });
        if (data && data.active == true) {
            const valid = await bcrypt.compare(req.body.pass, data.pass);
            if (valid) {
                const token = await data.gentoken();

                await userModel.updateOne({ email: req.body.email }, { $set: { token: token } }, { new: true });

                res.status(200).json({
                    status: true,
                    message: "Signin successfully..!",
                    data: token,
                });

            } else {
                res.status(400).json({
                    status: false,
                    message: "Your detail does not match..!",
                });
            }
        } else {
            res.status(400).json({
                status: false,
                message: "Your account does not exist, Please signup first..!",
            });
        }
    } catch (error) {
        res.status(400).json({
            status: false,
            message: error.message,
        });
    }
};

File: python-services/Controllers/user_controller.py
from Models.user_model import UserSignup, UserSignin
from Services.auth_service import hash_password, verify_password, create_access_token
from datetime import datetime, timezone
from fastapi.responses import JSONResponse
from fastapi import Request

async def signup(user: UserSignup, request: Request):
    users_collection = request.app.mongodb["users"]

    # Check if user already exists
    existing_user = await users_collection.find_one({"email": user.email})
    if existing_user:
        if not existing_user.get("active", True):
            # Reactivate user
            hashed_password = hash_password(user.password)

            token = create_access_token({"_id": str(existing_user["_id"])})

            await users_collection.update_one(
                {"email": user.email},
                {
                    "$set": {
                        "fname": user.fname,
                        "lname": user.lname,
                        "password": hashed_password,
                        "active": True,
                        "token": token,
                        "updated_at": datetime.now(timezone.utc),
                    }
                },
            )
            return JSONResponse(
                status_code=200,
                content={"status": True, "message": "Signup successful!", "data": token},
            )
        return JSONResponse(
            status_code=400,
            content={"status": False, "message": "User already exists!"},
        )

    # Create a new user
    hashed_password = hash_password(user.password)

    token = create_access_token({"email": user.email})

    new_user = {
        "fname": user.fname,
        "lname": user.lname,
        "email": user.email,
        "password": hashed_password,
        "token": token,
        "active": user.active,
        "created_at": datetime.now(timezone.utc),
        "updated_at": datetime.now(timezone.utc),
    }
    await users_collection.insert_one(new_user)
    return JSONResponse(
        status_code=201,
        content={"status": True, "message": "Signup successful!", "data": token},
    )


async def signin(user: UserSignin, request: Request):
    users_collection = request.app.mongodb["users"]

    # Find user by email
    existing_user = await users_collection.find_one({"email": user.email})
    if not existing_user or not existing_user.get("active", True):
        return JSONResponse(
            status_code=400,
            content={"status": False, "message": "Account does not exist. Please signup."},
        )

    # Verify password
    if not verify_password(user.password, existing_user["password"]):
        return JSONResponse(
            status_code=400,
            content={"status": False, "message": "Your details do not match!"},
        )

    # Generate token
    token = create_access_token({"_id": str(existing_user["_id"])})
    await users_collection.update_one(
        {"email": user.email},
        {"$set": {"token": token, "updated_at": datetime.now(timezone.utc)}},
    )
    return JSONResponse(
        status_code=200,
        content={"status": True, "message": "Signin successful!", "data": token},
    )
